File: src/commands/shared/bytes.ts
import type { CommandResult } from '../../types.js';
import { err } from '../../types.js';
import { parentPath } from '../../utils.js';
import type { CommandContext } from '../core.js';
import { materializedLimitError } from './io.js';

export type BytesInputResult =
  | { data: Uint8Array; source: string; error?: never }
  | { error: CommandResult; data?: never; source?: never };

export async function readBytesFromFileOrStdin(
  ctx: CommandContext,
  filePath: string | undefined,
  stdin: Uint8Array,
  commandName: string,
): Promise<BytesInputResult> {
  if (filePath !== undefined) {
    const info = await ctx.vfs.stat(filePath);
    const normalized = ctx.vfs.normalize(filePath);

    if (!info.exists) {
      return {
        error: err(`${commandName}: file not found: ${normalized}. Use: ls / or ls ${parentPath(filePath)}`),
      };
    }
    if (info.isDir) {
      return {
        error: err(`${commandName}: path is a directory: ${normalized}. Use: ls ${normalized}`),
      };
    }

    const limitError = materializedLimitError(ctx, commandName, normalized, info.size);
    if (limitError !== undefined) {
      return { error: limitError };
    }

    return {
      data: await ctx.vfs.readBytes(filePath),
      source: normalized,
    };
  }

  if (stdin.length > 0) {
    const limitError = materializedLimitError(ctx, commandName, 'stdin', stdin.length);
    if (limitError !== undefined) {
      return { error: limitError };
    }

    return {
      data: new Uint8Array(stdin),
      source: 'stdin',
    };
  }

  return {
    error: err(`${commandName}: no input provided. Usage: see 'help ${commandName}'`),
  };
}

File: python/src/onetool/commands/shared/bytes.py
from __future__ import annotations

from dataclasses import dataclass

from ...types import CommandResult, err
from ...utils import parent_path
from ..core import CommandContext
from .io import materialized_limit_error


@dataclass(frozen=True, slots=True)
class BytesInputResult:
    data: bytes | None = None
    source: str | None = None
    error: CommandResult | None = None


async def read_bytes_from_file_or_stdin(
    ctx: CommandContext,
    file_path: str | None,
    stdin: bytes,
    command_name: str,
) -> BytesInputResult:
    if file_path is not None:
        info = await ctx.vfs.stat(file_path)
        normalized = ctx.vfs.normalize(file_path)
        if not info.exists:
            return BytesInputResult(
                error=err(
                    f"{command_name}: file not found: {normalized}. "
                    f"Use: ls / or ls {parent_path(file_path)}"
                )
            )
        if info.is_dir:
            return BytesInputResult(
                error=err(f"{command_name}: path is a directory: {normalized}. Use: ls {normalized}")
            )

        limit_error = materialized_limit_error(ctx, command_name, normalized, info.size)
        if limit_error is not None:
            return BytesInputResult(error=limit_error)

        return BytesInputResult(data=await ctx.vfs.read_bytes(file_path), source=normalized)

    if stdin:
        limit_error = materialized_limit_error(ctx, command_name, "stdin", len(stdin))
        if limit_error is not None:
            return BytesInputResult(error=limit_error)
        return BytesInputResult(data=bytes(stdin), source="stdin")

    return BytesInputResult(error=err(f"{command_name}: no input provided. Usage: see 'help {command_name}'"))
